### PR TITLE
Convert components.js to typescript

### PR DESCRIPTION
**Pull request by MarkusJx**
This will convert ``static/components.js`` to typescript (see #2981)

Lots of types are still missing as there are still lots of typescript definitions missing.

### CODE REVIEW

**Comment by MarkusJx** [comment]
@supergrecko could you help me with the missing types and maybe take a look at the existing ones I assumed from the code usages (some seem a little weird like ``getOptViewWith``s param ``editorid`` is of type ``boolean``).

Any method/interface that has still missing types is marked with a ``TODO`` comment.

**Comment by supergrecko** [comment]
Could you rebase against main to include https://github.com/compiler-explorer/compiler-explorer/commit/896cf2f9cafc35e82de38fd94e6e6b72639636ea?

**Comment by mattgodbolt** [comment]
Hey @matsjla - do you think we're in a better spot to look at this now?

**Comment by matsjla** [comment]
> Hey @matsjla - do you think we're in a better spot to look at this now?

I think there's still a bunch of bits missing to do this properly, but I can put some extra effort into moving towards this over the next days/week

**Comment by mattgodbolt** [comment]
If @matsjla doesn't have time, I now feel like I have enough context to review. Some stuff has changed, so apologies for the merge conflict :(

**Comment by matsjla** [comment]
Sorry this took so long, but I've now got it to a point where I'm happy with integrating it. It still needs some more work here and there, but I think this is sufficient for this PR

File: static/components.interfaces.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+export interface ComponentConfig<S> {
```
Comment (mattgodbolt) [inline static/components.interfaces.ts:25]: Thank you for the accurate modeling of the "interfaces" (such as they are) here. I imagine it was a pain to reverse engineer the complexities that have grown over the years.

File: static/components.interfaces.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+type StateWithLanguage = {lang: string};
+type StateWithEditor = {source: number};
+type StateWithTree = {tree: number};
+type StateWithId = {id: number};
+type EmptyState = Record<never, never>;
+
+export type EmptyCompilerState = StateWithLanguage & StateWithEditor;
+export type PopulatedCompilerState = StateWithEditor & {
+    filters: Record<string, boolean>; // TODO: compilation filters
```
Comment (mattgodbolt) [inline static/components.interfaces.ts:39]: Thanks for leaving a TODO here; best to get something out that can be improved upon than obsessing over the minutiae. Thanks!

File: static/components.ts [hunk 3]
```diff
@@ -0,0 +1,866 @@
...
+    PopulatedGnatDebugViewState,
+    EmptyRustMacroExpViewState,
+    PopulatedRustMacroExpViewState,
+    EmptyRustHirViewState,
+    PopulatedRustHirViewState,
+    EmptyDeviceViewState,
+    PopulatedDeviceViewState,
+} from './components.interfaces';
+
+export const COMPILER_COMPONENT_NAME = 'compiler';
```
Comment (mattgodbolt) [inline static/components.ts:78]: I defer to @matsjla et al; I'm still learning here - but why these exported `CONSTANT_NAMES` here and not in the interfaces?

File: static/components.ts [hunk 4]
```diff
@@ -0,0 +1,866 @@
...
+            source: editorId,
+            lang,
+        },
+    };
+}
+
+/**
+ * Get a compiler component with the given configuration.
+ *
+ * We have legacy calls in the codebase (to keep shortlinks permanent) that use this function without
```
Comment (mattgodbolt) [inline static/components.ts:120]: Thanks for the explanatory comments!

File: static/hub.ts [hunk 5]
```diff
@@ -79,50 +79,41 @@ export class Hub {
...
-        layout.registerComponent(Components.getDiffView().componentName, (c, s) => this.diffFactory(c, s));
-        layout.registerComponent(Components.getOptView().componentName, (c, s) => this.optViewFactory(c, s));
-        layout.registerComponent(Components.getFlagsView().componentName, (c, s) => this.flagsViewFactory(c, s));
-        layout.registerComponent(Components.getPpView().componentName, (c, s) => this.ppViewFactory(c, s));
-        layout.registerComponent(Components.getAstView().componentName, (c, s) => this.astViewFactory(c, s));
-        layout.registerComponent(Components.getIrView().componentName, (c, s) => this.irViewFactory(c, s));
-        layout.registerComponent(Components.getDeviceView().componentName, (c, s) => this.deviceViewFactory(c, s));
-        layout.registerComponent(Components.getRustMirView().componentName, (c, s) => this.rustMirViewFactory(c, s));
-        layout.registerComponent(Components.getHaskellCoreView().componentName, (c, s) =>
+        layout.registerComponent(Components.EDITOR_COMPONENT_NAME, (c, s) => this.codeEditorFactory(c, s));
```
Comment (mattgodbolt) [inline static/hub.ts:82]: Much nicer way of rendezvousing on the name. Thanks.

File: static/components.ts (hunk 3 above)
Comment (matsjla) [reply static/components.ts:78]: Good catch, I'll move them! Thanks

File: static/components.interfaces.ts (hunk 2 above)
Comment (partouf) [reply static/components.interfaces.ts:39]: filters are defined here https://github.com/compiler-explorer/compiler-explorer/blob/main/types/features/filters.interfaces.ts#L25 - though not sure if they are equal to the ones in our UI

File: static/components.interfaces.ts [hunk 6]
```diff
@@ -0,0 +1,249 @@
+// Copyright (c) 2021, Compiler Explorer Authors
```
[the remark this answers is not in the document]
Comment (partouf) [reply static/components.interfaces.ts:1]: it is a PR from 2021 ;)

**Pull request merged**